typed-transitions/src/Protocol/ChainSync/Type.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | The type of the chain synchronisation protocol.
+--
+-- Since we are using a typed protocol framework this is in some sense /the/
+-- definition of the protocol: what is allowed and what is not allowed.
+--
+module Protocol.ChainSync.Type where
+
+import Protocol.Core
+
+
+-- | The states in the state transition diagram of the chain sync protocol.
+--
+data ChainSyncState where
+
+  -- | Both client and server are idle. The client can send a request and
+  -- the server is waiting for a request.
+  StIdle      :: ChainSyncState
+
+  -- | The client has sent a next update request. The client is now waiting
+  -- for a response, and the server is busy getting ready to send a response.
+  -- There are two possibilities here, since the server can send a reply
+  -- immediately or it can send an initial await message followed later by
+  -- the normal reply.
+  StNext      :: StNextKind -> ChainSyncState
+
+  -- | The client has sent an intersection request. The client is now waiting
+  -- for a response, and the server is busy getting ready to send a response.
+  StIntersect :: ChainSyncState
+
+  -- | Both the client and server are in the terminal state. They're done.
+  StDone      :: ChainSyncState
+  --TODO: currently there is no termination mechanism
+
+-- | Sub-cases of the 'StNext' state. This is needed since the server can
+-- either send one reply back, or two.
+--
+data StNextKind where
+  StCanAwait  :: StNextKind -- ^ The server can reply or send an await msg.
+  StMustReply :: StNextKind -- ^ The server must now reply, having already
+                                -- sent an await message.
+
+
+-- | A type to identify the client\/server partition of states in our protocol.
+data ChainSyncProtocol
+
+-- | We have to explain to the framework what our states mean, in terms of
+-- which party has agency in each state.
+--
+type instance Partition ChainSyncProtocol st client server terminal =
+  ChainSyncStatePartition st client server terminal
+
+-- | Idle states are where it is for the client to send a message,
+-- busy states are where the server is expected to send a reply.
+--
+type family ChainSyncStatePartition st client server terminal where
+  ChainSyncStatePartition  StIdle      client server terminal = client
+  ChainSyncStatePartition (StNext k)   client server terminal = server
+  ChainSyncStatePartition  StIntersect client server terminal = server
+  ChainSyncStatePartition  StDone      client server terminal = terminal
+
+
+-- | The messages in the chain sync protocol.
+--
+-- In this protocol the consumer always initiates things and the producer
+-- replies.
+--
+data ChainSyncMessage header point from to where
```

Comment:
Supposing there's a fork where the new chain contains `n > 1` blocks not in the old one. The producers says roll-back to the intersection point, then what happens next? A sequence of `n` roll-forwards? Each one giving the same tip point?

I think it's important that the producer can respond to intersect requests with a fork or extension. In the protocol here, the only way to realize that the chain has changed is to ask for the next header. But aren't there legitimate cases in which the consumer is not interested in downloading the producer's current chain, and instead wants to wait for it to change?

I also don't see the point of `MsgAwaitReply`. Is this needed for some timeout negotiation? Why not just block until the reply is ready and then send it?

Reply:
This protocol doesn't have the nice property that I wanted: the consumer never has to make a decision without knowing the tip and the read pointer. In this one, it only gets the tip after requesting next.

Reply:
In @dcoutts protocol, after roll-back to the intersection point the consumer receives `n` roll-forwards with successive points where the final one will be the tip (if it hasn't changed).

> But aren't there legitimate cases in which the consumer is not interested in downloading the producer's current chain, and instead wants to wait for it to change?

At least in this protocol the client has no evidence that it should wait.  This protocol just copies the producer chain to the client.

> This protocol doesn't have the nice property that I wanted: the consumer never has to make a decision without knowing the tip and the read pointer.

That's indeed a nice property to have, though using it might make the design a bit more complex.  The client would be able to make a decision knowing history of forks from other chain following protocols running on the same node (i.e. maybe some node already switched from that fork to a better one) or just comparing with its current chain.

Reply:
> I also don't see the point of MsgAwaitReply. Is this needed for some timeout negotiation?

Indeed, and it can be removed when we'll have keep alive messages.

Reply:
> That's indeed a nice property to have, though using it might make the design a bit more complex. The client would be able to make a decision knowing history of forks from other chain following protocols running on the same node (i.e. maybe some node already switched from that fork to a better one) or just comparing with its current chain.

The design of the _protocol_ is only marginally more complex (it's right here in this repository).

The design of the application of the protocol _could_ become more complex, but does not have to be. That the protocol _allows_ for the added complexity is good, though.